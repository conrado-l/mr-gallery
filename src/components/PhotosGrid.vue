<template>
  <div>
    <!-- Coolbox photo library -->
    <CoolLightBox
      :items="photos"
      :index="currentOpenedPhotoIndex"
      :useZoomBar="true"
      :fullScreen="true"
      :loop="false"
      :gallery="false"
      effect="fade"
      @close="currentOpenedPhotoIndex = null"
      @on-change="onPhotoDetailChange($event)"
    >
    </CoolLightBox>
    <!---->

    <!-- Infinity scroll -->
    <div v-infinite-scroll="onInfiniteScrollLoadMore"
         :infinite-scroll-distance="10">
      <v-row>
        <v-col
          v-for="(photo, photoIndex) in photos"
          :key="photo.id"
          class="d-flex child-flex"
          xs="1"
          sm="1"
          md="4"
        >
          <!-- Thumbnail photos -->
          <v-tooltip bottom>
            <template v-slot:activator="{ on, attrs }">
              <v-img
                class="grey lighten-2 clickable"
                :src="photo.thumb"
                v-ripple
                v-bind="attrs"
                v-on="on"
                data-test="thumbnail-photo"
                @click="onPhotoClick(photoIndex, photo.id)"
              >
                <template v-slot:placeholder>
                  <v-row
                    class="fill-height ma-0"
                    align="center"
                    justify="center"
                  >
                    <v-progress-circular
                      indeterminate
                      color="grey lighten-5"
                    ></v-progress-circular>
                  </v-row>
                </template>
              </v-img>
            </template>
            <span>Click for more details</span>
          </v-tooltip>
          <!---->
        </v-col>
      </v-row>
    </div>
    <!----->
  </div>
</template>

<script>
// Plugins
import CoolLightBox from 'vue-cool-lightbox'
import 'vue-cool-lightbox/dist/vue-cool-lightbox.min.css'
import infiniteScroll from 'vue-infinite-scroll'

/**
 * A grid of clickable photos, using infinity scroll, with the ability to open them in full screen, for more details
 **/
export default {

  name: 'PhotosGrid',
  props: {
    /** Contains the photos to be displayed **/
    photos: {
      type: Array,
      required: true
    },
    /** Checks if the photos are being fetched **/
    fetchingPhotos: {
      type: Boolean,
      required: true
    },
    /** Checks if the photo's details are being fetched **/
    fetchingPhotoDetails: {
      type: Boolean,
      required: true
    }
  },
  data: function () {
    return {
      currentOpenedPhotoIndex: null
    }
  },
  components: {
    CoolLightBox
  },
  directives: { infiniteScroll },
  methods: {
    /**
     * Notifies that the user scrolled down for loading more images
     /**/
    onInfiniteScrollLoadMore () {
      if (this.fetchingPhotos) {
        return
      }

      this.$emit('load-more-photos')
    },
    /**
     * Notifies that the user changed the photo from the detail modal
     * @param {number} photoIndex
     **/
    onPhotoDetailChange (photoIndex) {
      if (this.fetchingPhotoDetails) {
        return
      }

      this.notifyPhotoDetail(this.photos[photoIndex].id)
    },
    /**
     * Notifies that a photo was clicked, it also opens the detail modal for showing the photo in full size
     * @param {number} photoIndex
     * @param {string} photoId
     **/
    onPhotoClick (photoIndex, photoId) {
      this.notifyPhotoDetail(photoId)
      this.currentOpenedPhotoIndex = photoIndex
    },
    /**
     * Notifies that we need the details for a particular photo
     * @param {string} photoId
     */
    notifyPhotoDetail (photoId) {
      this.$emit('load-photo-detail', photoId)
    }
  }
}
</script>

<style lang="scss" scoped>
.clickable {
  cursor: pointer;
}
</style>
