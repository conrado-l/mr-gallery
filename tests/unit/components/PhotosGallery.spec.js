// Vue Test Utils
import { createLocalVue, mount } from '@vue/test-utils'

// Plugins
import Vuetify from 'vuetify'

// Components
import PhotosGrid from '@/components/PhotosGrid'

describe('PhotosGrid.vue', () => {
  const localVue = createLocalVue()
  let vuetify

  // Default props
  const defaultProps = {
    photos: [],
    fetchingPhotos: false,
    fetchingPhotoDetails: false
  }

  // Reusable factory mount
  const factoryMount = (props) => {
    return mount(PhotosGrid, {
      localVue,
      vuetify,
      propsData: {
        ...props
      }
    })
  }

  beforeEach(() => {
    vuetify = new Vuetify()
  })

  it('should render the component correctly', () => {
    const wrapper = factoryMount(defaultProps)

    expect(wrapper.element).toMatchSnapshot()
  })

  // TODO: find out why vm.$nextTick is not waiting for the VImg component to load the image, resulting in empty "src"
  // NOTE: the test runs properly when :lazy-src="photo.thumb" is set to the VImg component inside PhotosGrid.vue
  it('should render 10 thumbnail photos with the correct source URLs', async () => {
    // Photos mock array
    const photos = [{
      id: '1a5e86953ad5ac438130',
      thumb: 'http://interview.agileengine.com/pictures/cropped/0002.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/0002.jpg'
    }, {
      id: '49aa54843eaf3a0a49be',
      thumb: 'http://interview.agileengine.com/pictures/cropped/0015.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/0015.jpg'
    }, {
      id: 'd0d3fb4d11aaee9accce',
      thumb: 'http://interview.agileengine.com/pictures/cropped/0019.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/0019.jpg'
    }, {
      id: '5a83b798d52b0f673f66',
      thumb: 'http://interview.agileengine.com/pictures/cropped/0020.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/0020.jpg'
    }, {
      id: '305d191d6cb0c74159f3',
      thumb: 'http://interview.agileengine.com/pictures/cropped/005.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/005.jpg'
    }, {
      id: '155c9d2210b2bad61c1f',
      thumb: 'http://interview.agileengine.com/pictures/cropped/01.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/01.jpg'
    }, {
      id: 'cff3db4ea0c2809ed1e6',
      thumb: 'http://interview.agileengine.com/pictures/cropped/011129columbus1-01.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/011129columbus1-01.jpg'
    }, {
      id: '7319d1bd1e8c7614ac8f',
      thumb: 'http://interview.agileengine.com/pictures/cropped/018086b.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/018086b.jpg'
    }, {
      id: 'c0d2f9fe5f35132bd336',
      thumb: 'http://interview.agileengine.com/pictures/cropped/01sc087.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/01sc087.jpg'
    }, {
      id: 'e13a844e87c749edd2fc',
      thumb: 'http://interview.agileengine.com/pictures/cropped/02sc003.jpg',
      src: 'http://interview.agileengine.com/pictures/cropped/02sc003.jpg'
    }]

    // Instantiate the component
    const wrapper = factoryMount({
      ...defaultProps,
      photos
    })

    // Wait for the next tick
    await wrapper.vm.$nextTick()

    // Find the thumbnail photos
    const thumbnailPhotoWrappers = wrapper.findAll('.v-image__image.v-image__image--cover')

    // Check if 10 thumbnail photos were rendered
    expect(thumbnailPhotoWrappers.length).toEqual(10)

    // Check if the thumbnail photos sources are correct (VImg uses the "background-image" CSS property)
    thumbnailPhotoWrappers.wrappers.forEach((thumbnailPhoto, index) => {
      expect(thumbnailPhoto.element.style.backgroundImage).toBe(`url ("${photos[index].src}")`)
    })
  })
})
