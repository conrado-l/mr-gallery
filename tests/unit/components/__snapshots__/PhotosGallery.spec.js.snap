// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotosGrid.vue should render the component correctly 1`] = `
<div>
  <transition-stub
    name="cool-lightbox-modal"
  >
    <!---->
  </transition-stub>
   
  <div
    infinite-scroll-distance="10"
  >
    <div
      class="row"
    />
  </div>
</div>
`;
